File: legacy-nodejs/src/services/category.service.ts
import { ICategoryRepository } from '../repositories/interfaces';
import { CategoryRecord } from '../types/domain';
import { ConflictError, NotFoundError, ValidationError } from '../errors/AppError';
import { createCategorySchema } from '../validators/product.validators';

export class CategoryService {
  constructor(private readonly categoryRepository: ICategoryRepository) {}

  async listCategories(): Promise<CategoryRecord[]> {
    return this.categoryRepository.findAll();
  }

  async createCategory(input: { name: string }): Promise<CategoryRecord> {
    const parsed = createCategorySchema.safeParse(input);
    if (!parsed.success) {
      throw new ValidationError('Invalid category data', parsed.error.flatten());
    }

    const existing = await this.categoryRepository.findByName(parsed.data.name);
    if (existing) {
      throw new ConflictError('A category with this name already exists');
    }

    return this.categoryRepository.create(parsed.data.name);
  }

  /**
   * Category deletion is blocked if any active product still references it
   * (Milestone 1 §11 — "Category deletion blocked if products reference it").
   */
  async deleteCategory(id: string): Promise<void> {
    const category = await this.categoryRepository.findById(id);
    if (!category) {
      throw new NotFoundError('Category');
    }

    const hasProducts = await this.categoryRepository.hasProducts(id);
    if (hasProducts) {
      throw new ConflictError('Cannot delete a category that still has active products');
    }

    await this.categoryRepository.delete(id);
  }
}

File: src/Services/Implementations/CategoryService.cs
using ECommerceApi.DTOs.Categories;
using ECommerceApi.Exceptions;
using ECommerceApi.Repositories.Interfaces;
using ECommerceApi.Services.Interfaces;

namespace ECommerceApi.Services.Implementations;

public class CategoryService : ICategoryService
{
    private readonly ICategoryRepository _categoryRepository;

    public CategoryService(ICategoryRepository categoryRepository)
    {
        _categoryRepository = categoryRepository;
    }

    public async Task<List<CategoryResponse>> ListCategoriesAsync()
    {
        var categories = await _categoryRepository.FindAllAsync();
        return categories.Select(c => new CategoryResponse { Id = c.Id, Name = c.Name }).ToList();
    }

    public async Task<CategoryResponse> CreateCategoryAsync(CreateCategoryRequest request)
    {
        if (string.IsNullOrWhiteSpace(request.Name))
        {
            throw new ValidationException("Category name is required");
        }

        var existing = await _categoryRepository.FindByNameAsync(request.Name);
        if (existing is not null)
        {
            throw new ConflictAppException("A category with this name already exists");
        }

        var category = await _categoryRepository.CreateAsync(request.Name);
        return new CategoryResponse { Id = category.Id, Name = category.Name };
    }

    
    // Category deletion is blocked if any active product still references it
  
    public async Task DeleteCategoryAsync(Guid id)
    {
        var category = await _categoryRepository.FindByIdAsync(id);
        if (category is null)
        {
            throw new NotFoundAppException("Category");
        }

        var hasProducts = await _categoryRepository.HasActiveProductsAsync(id);
        if (hasProducts)
        {
            throw new ConflictAppException("Cannot delete a category that still has active products");
        }

        await _categoryRepository.DeleteAsync(id);
    }
}
